### PR TITLE
Add Apple Provisioning profile

### LINKED ISSUE #1434: provide config profiles for OS X/macOS and iOS
OS X/macOS and iOS support configuration profiles. This allows extremely easy setup of CalDAV and CardDAV clients.

### PR DESCRIPTION
**Pull request by georgehrke**
fixes #1434 

Todo:
- [x] create provisioning root collection and add apple provisioning node to it
- [x] tests

### CODE REVIEW

**Comment by georgehrke** [comment]
It seems that even with the most detailed provisioning profile iOS / macOS will still look for the .well-known/caldav  / .well-known/carddav file :/ 

https://netfuture.ch/2013/07/dealing-with-apple-mobileconfig-configuration-profiles-quirks/

@MorrisJobke @rullzer Did we fix that issue that .well-known is only installed inside the Nextcloud-folder but not in the domains root?

**Comment by MorrisJobke** [comment]
We did not, because it then needs to be configured in the Webserver itself

**Comment by jospoortvliet** [comment]
So that can't be 100% fixed? :sob:

**Comment by MorrisJobke** [comment]
> @MorrisJobke @rullzer Did we fix that issue that .well-known is only installed inside the Nextcloud-folder but not in the domains root?

We have a setup check for this nowadays. So I would assume that it will work for now and provide it.

**Comment by MorrisJobke** [comment]
I rebased it. I would say, we should continue this and can improve in the future iteratively.

**Comment by MorrisJobke** [comment]
@georgehrke Looks like this one will not make it in time, right? Would like to have it merged by tomorrow evening.

**Comment by rullzer** [comment]
Lets rebase and get this in. Yeah it won't work in 100% of the cases but so be it there is no way to do that.

**Comment by georgehrke** [comment]
We should remove `CalDAVPrincipalURL` and `CardDAVPrincipalURL`, because they are not used anymore. 

Will rebase, fix that and add tests next week.

**Comment by MorrisJobke** [comment]
> Will rebase, fix that and add tests next week.

Could you have a look at this? So we can finally merge it :)

**Comment by georgehrke** [comment]
tl;dr of what changed:
- [x] added unit tests
- [x] enforce HTTPs. Using the profile on HTTP silently fails.
- [x] removed principalURI and require admins to properly configure well-known.

**Comment by georgehrke** [comment]
please review @nextcloud/mac

**Comment by rullzer** [comment]
I can't test since.. well no mac...

@georgehrke for a followup PR what would be nice

1. Add a check for the well-known
2. On the groupware page detect if a user is on a Apple device and show a link

**Review by rullzer — APPROVED** [review]
Looks good from my side. And tests :tada: 
But I can't tests since no macOS/iOS. so. :elephant:

**Comment by marinofaggiana** [comment]
> Looks good from my side. And tests 🎉
> But I can't tests since no macOS/iOS. so. 🐘

At end of next week I prepare a test for iOS

cc @georgehrke

File: apps/dav/lib/Provisioning/Apple/AppleProvisioningPlugin.php [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+	 *
+	 * @param IUserSession $userSession
+	 * @param IURLGenerator $urlGenerator
+	 * @param ThemingDefaults $themingDefaults
+	 * @param IRequest $request
+	 * @param IL10N $l10n
+	 * @param \closure $uuidClosure
+	 */
+	public function __construct(IUserSession $userSession, IURLGenerator $urlGenerator,
+								ThemingDefaults $themingDefaults, IRequest $request,
```
Comment (MorrisJobke) [inline apps/dav/lib/Provisioning/Apple/AppleProvisioningPlugin.php]: This will fail if the theming app is disabled. Let me check if I can fix this.

**Review by MorrisJobke — APPROVED** [review]
Tested and works really nice 👍
